tt_metal/tt_stl/overloaded.hpp

```diff
@@ -0,0 +1,38 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+namespace tt::stl {
+
+// `overloaded` allows to combine multiple lambdas into a single object with the overloaded `operator()`.
+// This is useful for creating visitor objects for using in `std::visit`, for example:
+//
+// void f(const std::variant<int, std::string>& v) {
+//   std::visit(overloaded {
+//     [](int i) { std::cout << "Got an int: " << i << std::endl; },
+//     [](const std::string& str) { std::cout << "Got a string: " << str << std::endl; },
+//   }, v);
+// }
+//
+//
+// Generic lambdas can also be used to simplify handling of variant types:
+//
+// void f(const std::variant<std::vector<int>, int, std::string>& v) {
+//   std::visit(overloaded {
+//     [](const std::vector<int>& vec) { std::cout << "Got a vector of size: " << vec.size() << std::endl; },
+//     [](const auto& other) { std::cout << "Got something else: " << other << std::endl; },
+//   }, v);
+// }
+//
+//
+//
+template <typename... Ts>
+struct overloaded : Ts... {
+    using Ts::operator()...;
+};
+template <typename... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
```

Comment:
You might consider supporting non-copyable lvalue references as well:
(C++17)
```cpp
namespace detail {

template <class T, class = void>
struct overloaded_base {
  using type = T;
};

template <class T>
struct overloaded_base<T &, std::enable_if_t<not std::is_copy_constructible_v<T>>> {
  using type = std::reference_wrapper<T>;
};

template <class T>
using overloaded_base_t = typename overloaded_base<T>::type;

}  // namespace detail

template <typename... Ts>
struct overloaded : detail::overloaded_base_t<Ts>... {
    using Ts::operator()...;
};

template <typename... Ts>
overloaded(Ts &&...) -> overloaded<Ts...>;
```
(or C++20)
```cpp
namespace detail {

template <class T>
struct overloaded_base {
  using type = T;
};

template <class T>
  requires (not std::copy_constructible<T>)
struct overloaded_base<T &> {
  using type = std::reference_wrapper<T>;
};

template <class T>
using overloaded_base_t = typename overloaded_base<T>::type;

}  // namespace detail

template <typename... Ts>
struct overloaded : detail::overloaded_base_t<Ts>... {
    using Ts::operator()...;
};

template <typename... Ts>
overloaded(Ts &&...) -> overloaded<Ts...>;
```

Reply:
For a concrete use-case, note that this allows you to pass a named lambda with a capture containing a `std::unique_ptr`, without move-constructing the lambda into the `overloaded` instance.

Reply:
Thanks, makes sense! I used the sfinae version in case something happens w.r.t. cpp20 support, since this is intended to be used widely.